compiler/rustc_mir/src/interpret/cast.rs

```diff
@@ -269,12 +269,61 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     Immediate::new_slice(ptr, length.eval_usize(*self.tcx, self.param_env), self);
                 self.write_immediate(val, dest)
             }
-            (&ty::Dynamic(..), &ty::Dynamic(..)) => {
-                // For now, upcasts are limited to changes in marker
-                // traits, and hence never actually require an actual
-                // change to the vtable.
+            (&ty::Dynamic(ref data_a, ..), &ty::Dynamic(ref data_b, ..)) => {
                 let val = self.read_immediate(src)?;
-                self.write_immediate(*val, dest)
+                if data_a.principal_def_id() == data_b.principal_def_id() {
+                    self.write_immediate(*val, dest)
+                } else {
+                    // trait upcasting coercion
+                    let principal_a = data_a.principal().expect(
+                        "unsize_into_ptr: missing principal trait for trait upcasting coercion",
+                    );
+                    let principal_b = data_b.principal().expect(
+                        "unsize_into_ptr: missing principal trait for trait upcasting coercion",
+                    );
+
+                    let vptr_entry_idx = self.tcx.vtable_trait_upcasting_coercion_new_vptr_slot((
+                        principal_a.with_self_ty(*self.tcx, src_pointee_ty),
+                        principal_b.def_id(),
+                    ));
+
+                    if let Some(entry_idx) = vptr_entry_idx {
+                        let (old_data, old_vptr) = match *val {
+                            Immediate::ScalarPair(data, vptr) => (data, vptr),
+                            Immediate::Scalar(..) => span_bug!(
```

Comment:
Please don't `match` on `Immediate` here; instead add a method like `to_scalar_or_uninit_pair` or so to `Immediate` in `operand.rs`.

I am a bit surprised we don't have such a method yet; it's probably because so far all interactions with pointers to unsized types went through `MemPlace` (which has a `vtable()` method) -- but creating a `MemPlace` here would make little sense.

Reply:
In fact you can make it `to_scalar_pair` since you are doing `check_init` on both components anyway.

(This matches the existing `to_scalar_or_uninit` and `to_scalar` methods.)

Reply:
Done.